### ISSUE TITLE
cmd/compile: calling string(...) on generic slices fails regardless of type constraints.

### ISSUE THREAD

**Issue by tmr232** [golang/go#50421]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18beta1 windows/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes (1.18 beta)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
set GO111MODULE=
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\tamir.bahar\AppData\Local\go-build
set GOENV=C:\Users\tamir.bahar\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=C:\Users\tamir.bahar\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\Users\tamir.bahar\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=C:\Program Files\Go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLDIR=C:\Program Files\Go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.18beta1
set GCCGO=gccgo
set GOAMD64=v1
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set GOMOD=NUL
set GOWORK=
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -mthreads -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=C:\Users\TAMIR~1.BAH\AppData\Local\Temp\go-build3924345009=/tmp/go-build -gno-record-gcc-switches
</pre></details>

### What did you do?

Compile the following code: 
```
func toString[T byte|rune](slice []T) string {
   return string(slice)
}
```

### What did you expect to see?

Since `string(...)` can take either a `[]byte` or a `[]rune`, I expected the code to compile successfully.

### What did you see instead?

Compilation fails with:
```
cannot convert slice (variable of type []T) to type string
```

See https://gotipplay.golang.org/p/ZgvdU6Xv9jB

**Comment by ALTree** [comment]
cc @griesemer

**Comment by griesemer** [comment]
Pet the [current spec rules](https://tip.golang.org/ref/spec#Conversions) on generic conversions, this conversion is not permitted.

Note that the following code does work:
```Go
func toString[T []byte|[]rune](slice T) string {
        return string(slice)
}
```

Maybe the rules need to be relaxed.
@ianlancetaylor for additional input.

**Comment by tmr232** [comment]
Thanks!

As motivation for the rule relaxation, I'll present my use-case. 

I have a generic type `Iterator[T]` that represents a Python-like iterator (it can be any other collection that's generic over the element type), and I want to convert it to a string.

Due to the current rules, I cannot just write the following:
```go
func ToString[T rune | uint8](iter Iterator[T]) string {
	return string(ToSlice(iter))
}
```

Trying to work around it doesn't work either:
```go
func sliceToString[T []rune | []uint8](slice T) string {
	return string(slice)
}

func ToString[T rune | uint8](iter Iterator[T]) string {
	return sliceToString(ToSlice(iter))
}
```

And fails with
```
[]T does not implement []rune|[]uint8
```

Which is hard for me to understand as `T` is either `rune` or `uint8`, making `[]T` effectively `[]rune|[]uint8`.


This means that I'll have to write 2 `toString` functions. One for `rune` and one for `uint8`. 
It's not that bad in that one case, but I think it would also mean that code that is generic over string-elements would either need to be written twice (once for `uint8` and once for `rune`), or that it will need non-generic free-functions for converting back to string.

**Comment by ianlancetaylor** [comment]
Thanks.  We're not going to change this for 1.18.  In the long run I think we should probably permit this, but let's leave real consideration for later.  Thanks.

**Comment by go101** [comment]
> Per the current spec rules on generic conversions, this conversion is not permitted.

@griesemer Could you more specific on which line in the current spec disallows this conversion? I only found this line

> A non-constant value x can be converted to type T in any of these cases: 
> * x is an integer or a slice of bytes or runes and T is a string type. 

The line doesn't mention type parameters in this line.

**Comment by ianlancetaylor** [comment]
In your example `slice` is not a slice of bytes or runes, it is a slice of `T`, which is a type parameter.  The rule you mention ("x is an integer or a slice of bytes or runes and T is a string type") doesn't apply.  I'm being pedantic, but that is what the spec is: pedantic.

**Comment by go101** [comment]
Is this the same problem?
```Go
type Bytes []byte

func hos[T Bytes](x []byte) T {
	return x // okay
}

func lay[T []byte](x Bytes) T {
	return x // error
}

func rez[T []byte](x Bytes) T {
	return T(x) // okay
}
```

**Comment by griesemer** [comment]
Making this a proposal for language change.

**Comment by adonovan** [comment]
If core types are removed from the spec, would the type checking of the conversion `string(slice)` become legal because all elements of the type would support that operation? In other words, might this just fall out as a consequence of https://github.com/golang/go/issues/70128?

**Comment by griesemer** [comment]
I don't think so. Core types are not involved in a `string(slice)` conversion.

**Comment by adonovan** [comment]
> I don't think so. Core types are not involved in a `string(slice)` conversion.

Quite right, and sorry for my being too lazy to check the spec, which basically says (in rather too many words) that it checks the entire cross product of the two type sets for convertibility, so it is already "post-core-type".

FWIW, this variant of the original program, using a type parameter for the slice type, is legal:
```go
func toString[T []byte|[]rune](slice T) string {
   return string(slice)
}
```
and may be an acceptable workaround.

**Comment by kendfss** [comment]
FWIW, there's another layer to this issue worth considering: this conversion is
also forbidden in type switches, even when the compiler has narrowed the type to
valid options.

Example:

```go
func Convert(buf any) string {
	switch v := buf.(type) {
	case []rune, []byte:
		return string(v) // cannot convert v (variable of interface type any) to type string: need type assertion
	default:
		return fmt.Sprintf("%v", v)
	}
}
```

The simplest workaround requires duplicating the branches:

```go
func test(arg any) string {
	switch arg := arg.(type) {
	case []byte:
		return string(arg)
	case []rune:
		return string(arg)
	default:
		return fmt.Sprintf("%v", arg)
	}
}
```

The generic version requires unsafe:

```go
func test2[T ~rune | ~byte](arg []T) string {
	var t T
	switch any(t).(type) {
	case byte:
		return string(*(*[]byte)(unsafe.Pointer(&arg)))
	case rune:
		return string(*(*[]rune)(unsafe.Pointer(&arg)))
	}
	panic("unreachable")
}
```

This feels inconsistent with Go's simplicity goals. It's particularly confusing
because this works:

```go
func test3[T ~rune | ~byte | ~string](arg T) string {
	return string(arg)
}
```

But this doesn't:

```go
func test4(arg any) string {
	switch arg := arg.(type) {
	case rune, byte, string:
		return string(arg) // cannot convert arg (variable of interface type any) to type string: need type assertion
	default:
		return fmt.Sprintf("%v", arg)
	}
}
```

I think the compiler knows the type is valid in all these cases, but requires
either duplication or unsafe code to work around type system limitations.

**Comment by apparentlymart** [comment]
The observation in the previous comment about type switch arms not causing a type set that's a union of all of the types mentioned in `case` seems related to https://github.com/golang/go/issues/65031, in case that historical discussion is helpful context.
